ggml-cuda.cu

```diff
@@ -2432,7 +2521,71 @@ GGML_CALL static enum ggml_status ggml_backend_cuda_graph_compute(ggml_backend_t
         }
         GGML_ASSERT(ok);
     }
+    }
+
+    #ifdef USE_CUDA_GRAPH
+    if(useCudaGraph && (cudaGraphUpdateRequired)) { // End CUDA graph capture
+        CUDA_CHECK(cudaStreamEndCapture(cuda_ctx->stream(), &cudaGraph.graph));
+    }
+    if(useCudaGraph){
+
+        if(cudaGraph.instance == nullptr) { // Create executable graph from captured graph.
+            CUDA_CHECK(cudaGraphInstantiate(&cudaGraph.instance, cudaGraph.graph, NULL, NULL, 0));
+        }
+
+
+        // Perform update to graph (if required for this token), and change copy parameter (required for every token)
 
+        if(cudaGraphUpdateRequired) {
+            // Extract nodes from graph
+            if(cudaGraph.numNodes == 0) {
+                CUDA_CHECK(cudaGraphGetNodes(cudaGraph.graph, nullptr, &cudaGraph.numNodes));
+            }
+            CUDA_CHECK(cudaGraphGetNodes(cudaGraph.graph, cudaGraph.nodes, &cudaGraph.numNodes));
+
+            // Loop over nodes, and extract kernel parameters fro each node
+            for(size_t i=0; i<cudaGraph.numNodes; i++) {
+                cudaGraphNodeType nodeType;
+                CUDA_CHECK(cudaGraphNodeGetType(cudaGraph.nodes[i], &nodeType));
+                if (nodeType == cudaGraphNodeTypeKernel) {
+                    auto statRT = cudaGraphKernelNodeGetParams(cudaGraph.nodes[i], &cudaGraph.params[i]); // Get params using runtime
+                    if(statRT == cudaErrorInvalidDeviceFunction) {
+                        // Fails due to incorrect handling by CUDA runtime of CUDA BLAS node.
+                        // We don't need to update blas nodes, so clear error and move on.
+                        cudaGetLastError();
+                    }
+                }
+            }
+        }
+
+        // Update copy kernel param (required every token)
+        if(!cudaGraphUpdateRequired) { // on update steps, the live parameters will already be captured
+            int k=0;
+            for(size_t i=0; i<cudaGraph.numNodes; i++) {
+                if(cudaGraph.params[i].func == ggmlCudaCpyFn) {
+                    char** updatedKernelArgPointer = updatedKernelArg[k++];
+                    cudaGraph.params[i].kernelParams[1] = updatedKernelArgPointer;
+                    CUDA_CHECK(cudaGraphKernelNodeSetParams(cudaGraph.nodes[i], &cudaGraph.params[i]));
+                }
+            }
+        }
+
+        // Update graph executable
+        cudaGraphExecUpdateResultInfo resultInfo;
+        auto stat = cudaGraphExecUpdate(cudaGraph.instance, cudaGraph.graph, &resultInfo);
+        if(stat == cudaErrorGraphExecUpdateFailure)
+        {
```

Comment:
`{` should be on the same line as `if (...)`

Reply:
Done